salt/utils/jinja.py

```diff
@@ -196,7 +196,7 @@ def get_source(self, environment, template):
             }
             environment.globals.update(tpldata)
 
-        if _template in self.cached:
+        if _template in self.cached or os.path.exists(_template):
```

Comment:
Wouldn't it be better if we checked it was, in fact, an absolute path?
Just checking for the existence opens up for relative path security issues?

And, I'm currently on PTO so I can't check deeper, but, wouldn't absolute paths also open salt up for security issues?

Reply:
Responding from memory, if `_template` ends up in `self.cached`, that means that it's either a relative or absolute path _inside the salt filesystem_. So if it hits the `or` here, then it's _outside_ the salt filesystem. I don't think it's security-relevant whether the path is absolute or relative if we're deciding that we're allowing access outside the salt filesystem (retaining pre-existing functionality). However, I had a similar thought when I was writing the original patch... If we don't merge this PR, the change in functionality could be considered as fixing a security hole.

Does the Core Team want to take a poll and weigh the risk against breaking existing functionality?

Reply:
I would say lets label this as fixing a security hole. not sure rendering outside of the salt fileserver or at least the cache is really needed.

Reply:
Looping in @sabedevops and @supfors to speak to the use case.

Reply:
also @iaingeorgeson

Reply:
To answer the direct question, our use case is that we shard some configuration file templates. Some of the shard reside in an data sources with respect to Salt, and are dropped into place during system provisioning. The template that resides in the salt base serves as an interface to populate required variables and include that external template shard before final rendering.

As an aside, breaking the user experience in this way should always be last resort, in my opinion, and should only happen when the security concern is egregious and cannot be ignored. Fixing a security hole wasn't the goal of the initial patch, for example. Furthermore, I would question what security hole it is really fixing? The types of failures I'm imagining really reside in the process controls implemented by the user outside the scope of Salt. Can you explain what the concern is more concretely and how ensuring the template resides in the salt base addresses that concern? 

Now I'm thinking I have to emergency patch our workflow and the hundreds of salt-masters we deploy. If that is the case, I would really like to understand our exposure in the majority of our systems that precede 3005.

Reply:
I'm not sure that it's actually any kind of security hole. You can already do something like this:

```
/path/to/some/file:
    file.managed:
        - source: /path/to/literally/anywhere/on/the/filesystem
        - template: jinja
```

Unless I'm misunderstanding the path that results in this code being called. If we're talking about the above scenario, there's no security hole to be plugged -- it's just normal Salt functionality.

Reply:
(relative source paths I think also work there, but I'm not checking that right now)